[AndroidCrypto] Implement X509 chain building

elinor-fung opened the pull request:
There isn't a way to get rich status information (like what .NET APIs expose) or ignore certain validation around chain building on Android. In order to build a chain at all through the APIs available on Android, there must be a path with a valid time, signature, name constraints, policy constraints, and trusted root.

- Attempt to build the chain without extra validation (e.g. revocation checking) through [`CertPathBuilder`](https://developer.android.com/reference/java/security/cert/CertPathBuilder)
  - Per above, there is still a lot of validation that cannot be bypassed. if any of those can't be satisfied, building the chain simply fails with an exception - without anything more detailed than that a valid chain could not be constructed.
  - In these failure cases, `X509Chain.ChainElements` is empty and `X509Chain.ChainStatus` has `PartialChain` with whatever the exception message was.
- Do a validation pass over the built chain through [`CertPathValidator`](https://developer.android.com/reference/java/security/cert/CertPathValidator)
- Propagate exception message from Android cert path building APIs as chain status items
  - If we have an index (failure was in additional validation after the path was built), associate it with the corresponding chain element. Otherwise, just add it to the overall status.
  - If we have a failure reason (failure was in additional validation after the path was built, API level is 24+), map that to a corresponding status. Otherwise, report `PartialChain`.

Tests:
- Update `RevocationResponder` to handle POST requests for OCSP
- A bunch of the tests relied on `AllowUnknownCertificateAuthority` to validate other scenarios. On Android, you can't build a chain at all in that case.
  - For those tests, I made them use a custom trust store with a self-issued cert on Android so that they could build a chain and do whatever other validation the test was targeting.
- We don't have the detailed status information on Android, so a lot of the validation for invalid chains was changed to just check that it was invalid and had no elements.
- Skip tests for AIA fetching and offline mode of revocation

This gets `ChainTests`, `DynamicChainTests`, and `DynamicRevocationTests` (outerloop) working. There is one failure in `CertificateRequestChainTests`:
  - `CreateChain_RSAPSS` - fails to build any chain (zero-element, `PartialChain`)

Platform not supported:
- Ignoring verification that we can't bypass on Android: `AllowUnknownCertificateAuthority`, `IgnoreInvalidName`, `IgnoreInvalidPolicy`, `IgnoreTimeNotValid`
- Using a custom trust store with no (self-issued) certs
- Revocation check in `Offline` mode
  - Treated as `Online` 
- Revocation check with `EndCertificateOnly` when newer Android APIs supporting it aren't available
  - Treated as `ExcludeRoot`
- Revocation check with `EntireChain`
  - Treated as `ExcludeRoot`
- AIA fetching: https://docs.oracle.com/javase/8/docs/technotes/guides/security/certpath/CertPathProgGuide.html#AIA
  - This is disabled by default and requires setting a system property to enable, which we can't really do properly
  - Would probably require doing the manual downloading like what the Unix/OpenSSL implementation does

cc @jkoritzinsky @steveisok @AaronRobinsonMSFT @bartonjs

elinor-fung committed:
Basic chain building

TODO:
- set custom trust root
- handle revocation checking
- propagate errors/status where possible

elinor-fung committed:
Propagate error messages and set custom trust root certs

elinor-fung committed:
Map failure reason to chain status when possible

elinor-fung committed:
Share x509 enums

elinor-fung committed:
Update tests to reflect limited support on Android

elinor-fung committed:
Handle revocation when newer APIs are available

jkoritzinsky commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_jni.c:
```diff
@@ -349,14 +441,38 @@ void ReleaseLRef(JNIEnv *env, jobject lref)
+    jclass klass = NULL;
+    if (!TryGetClassGRef(env, name, &klass))
+    {
+        LOG_ERROR("optional class %s was not found", name);
```
```suggestion
        LOG_DEBUG("optional class %s was not found", name);
```

or 

```suggestion
        LOG_WARNING("optional class %s was not found", name);
```

jkoritzinsky commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -245,18 +254,28 @@ public static void SystemTrustCertificateWithCustomRootTrust(bool addCertificate
-                    Assert.Equal(X509ChainStatusFlags.UntrustedRoot, chain.AllStatusFlags());
+                    // Android does not support an empty custom root trust
+                    // Only self-issued certs are treated as trusted anchors, so building the chain
+                    // should through PNSE regardless of whether or not testCert is added to the store
```
```suggestion
                    // should throw PNSE regardless of whether or not testCert is added to the store
```

jkoritzinsky commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -175,7 +181,10 @@ public static void TestResetMethod()
                 Assert.False(valid);
                 chainHolder.DisposeChainElements();
 
+                // Make sure AllowUnknownCertificateAuthority is set, as this test we verifies it does not get reset
```
Can you reword this?

elinor-fung replied:
Just throwing random words together over here...

jkoritzinsky commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c:
```diff
@@ -0,0 +1,484 @@
+#include <stdbool.h>
+#include <string.h>
+
+#define INIT_LOCALS(name, ...) \
+    enum { __VA_ARGS__, count_##name }; \
+    jobject name[count_##name] = { 0 }; \
+
+#define RELEASE_LOCALS(name, env) \
+{ \
+    for (int i_##name = 0; i_##name < count_##name; ++i_##name) \
+    { \
+        jobject local = name[i_##name]; \
+        if (local != NULL) \
+            (*env)->DeleteLocalRef(env, local); \
+    } \
+} \
```
At this point do we want to unify all of the definitions of these macros we have into one definition in pal_jni?

jkoritzinsky reviewed (commented):
Not familiar enough with the cert side to fully grok if the logic is correct, but generally looks good to me. Just a few small nits from my first review pass.

elinor-fung commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                    {
+                        if ((flags & unsupported) == unsupported)
+                        {
+                            exception = new PlatformNotSupportedException(SR.Format(SR.Chain_VerificationFlagNotSupported, unsupported));
```
This is explicitly doing PNSE if some ignore flags are set that Android doesn't handle and the chain couldn't be built. We don't really have a way of determining why it couldn't be built - it might not actually be for one of the things the user tried to ignore. Maybe it would make sense to just let it return false without throwing? (and document that certain verification flags aren't supported)

bartonjs replied:
Hm. Throwing here does seem weird... because "you can use this value unless it maybe mattered" is weird.

Throwing always in chain.Build when an invalid flag is specified makes sense to me.  Or, yeah, returning false and not throwing.

The behavior for any of these cases is the zero-certificate PartialChain, right?  If so, just returning false is probably sufficient.

elinor-fung replied:
Yep, any of these would result in zero-certificate PartialChain. Will go with returning false.

elinor-fung committed:
Apply suggestions from code review

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

elinor-fung committed:
Move INIT_LOCALS, RELEASE_LOCALS, RELEASE_LOCALS_ENV macros to pal_jni

vcsjones commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
             bool disableAia)
         {
-            throw new NotImplementedException(nameof(BuildChain));
+            var chainPal = new AndroidCertPath();
```
It doesn't look like we do anything with the `disableAia` parameter here. This maps to `X509ChainPolicy.DisableCertificateDownloads` from the public API.

Does Android have a way to prevent fetching intermediate certificates? If not, should a PNSE be thrown if someone sets it to true? Should we ignore it and document that it does not work in Android?

There is some precedent for ignoring it with macOS. In macOS, this property only works if revocation checking is disabled as well.

elinor-fung replied:
It is actually that we don't have a good way to enable fetching. This is part of my 'not handled in this change'. By default, the fetching is disabled and can only be enabled though a [system property](https://docs.oracle.com/javase/8/docs/technotes/guides/security/certpath/CertPathProgGuide.html#AIA) which gets read and stored once. I don't think we really have a good way of dealing with that - I think we'd have to do something like the Unix/OpenSSl implementation's `CertificateAssetDownloader`.

bartonjs replied:
Until the need is proven otherwise, I'm OK with saying/believing that Android's use cases don't require AIA fetching.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.h:
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 #include "pal_jni.h"
+#include <pal_x509_types.h>
```
```suggestion
#include "pal_x509_types.h"
```

Quotes are local-first, right?

elinor-fung replied:
I had this because I put `pal_x509_types.h` is under Common, not in this folder?

bartonjs replied:
Ah, I'd missed that it was a different directory.  Nevermind.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c:
```diff
@@ -0,0 +1,470 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_x509chain.h"
+#include <pal_x509_types.h>
```
```suggestion
#include "pal_x509_types.h"
```

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs:
```diff
@@ -481,8 +535,16 @@ public static void NameConstraintViolation_ExcludedTree_Dns()
+                // the leading period. It checks expects the period as part of the left-side
+                // labels and not the constraint when doing validation.
+                // Use an excluded DNS name constraint without the period: example.com
+                nameConstraints = "3012A110300E820C6578616D706C652E636F6D";
```
The lengths in this value are invalid (they exceed the payload length).  Here's the corrected version:

```suggestion
                nameConstraints = "3011A10F300D820B6578616D706C652E636F6D";
```

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -326,7 +348,7 @@ public enum BuildChainCustomTrustStoreTestArguments : int
 
         [Fact]
-        public static void BuildChainWithSystemTrustAndCustomTrustCertificates()
+        public static void BuildChainWithSBystemTrustAndCustomTrustCertificates()
```
```suggestion
        public static void BuildChainWithSystemTrustAndCustomTrustCertificates()
```

I don't think we need the extra B :)

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -825,8 +873,8 @@ public static void ChainErrorsAtMultipleLayers()
------BEGIN CERTIFICATE-----
+            byte[] rootBytes = Encoding.ASCII.GetBytes(
+@"-----BEGIN CERTIFICATE-----
 MIICyjCCAbKgAwIBAgIIKKt3K3rRbvQwDQYJKoZIhvcNAQELBQAwFDESMBAGA1UEAxMJVGVzdCBS
```
Please revert these unnecessary style changes.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                X509Certificate2Collection customTrustStore,
+                X509ChainTrustMode trustMode)
+            {
+                List<IntPtr> extraCerts = new List<IntPtr>() { cert.Handle };
```
There's a bit of a race condition problem here with disposing.  A more robust solution would be to have a List<SafeX509Handle> that holds all of the Pal SafeHandle values, they get DangerousAddRef'd in the beginning, then DangerousRelease'd in a finally.  (Once they're all add-ref'd then make an array for the IntPtr values).

Alternatively, the P/Invoke marshaller might inherently support `SafeHandle[]` as doing the upref/downref thing and marshalling as the handle values.  (I don't remember if it does or not).

elinor-fung committed:
Apply suggestions from code review

Co-authored-by: Jeremy Barton <jbarton@microsoft.com>

elinor-fung committed:
Merge branch 'main' into chainPal-android

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                    // There is no way to bypass time, trusted root, name, or policy constraint
+                    // validation on Android. It will not build any chain without these all
+                    // being valid.
+                    X509VerificationFlags[] unsupportedFlags = new X509VerificationFlags[]
```
This should probably be a static readonly so it doesn't have to get recreated each time.  (`private static readonly X509VerificationFlags[] s_unsupportedVerificationFlags = ...`)

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                }
+
+                List<IntPtr> customTrustCerts = new List<IntPtr>();
+                bool useCustomRootTrust = trustMode == X509ChainTrustMode.CustomRootTrust;
```
There should be an assert here somewhere that if the mode isn't CustomRootTrust it's SystemTrust, so that the asserts help us track down where we need to support (or fail to support) when future modes are added.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                {
+                    if (revocationFlag == X509RevocationFlag.EntireChain)
+                    {
+                        throw new NotImplementedException($"{nameof(Evaluate)} (X509RevocationFlag.{revocationFlag})");
```
This is a weird error to emit to callers.

Probably just treat it as EntireChainIgnoreRoot, as that's the best-fit mapping.  In practice, roots don't get revoked (they have to revoke themselves, which is weird).

elinor-fung replied:
Would you rather treat it as EntireChainIgnoreRoot or actually throw PNSE?

bartonjs replied:
I'd rather treat it as EntireChainIgnoreRoot, since all that anyone will do is "if android, use EntireChainIgnoreRoot instead of EntireChain".  We can do that for them, and there's not much semantic difference.

If it was "EndCertificateOnly is the only mode we support" there's a major risk there, and we'd either need to fail, or work around the problem by "OK, so that chain passed... pop up a level and try again...".  (Besides, maybe we could get EntireChain by doing "fine, we'll build the chain again just for the root... happy?" but we'd want to observe that they don't just ignore revocation checks on self-issued certificates before committing that code)

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+
+                    if (!Interop.AndroidCrypto.X509ChainSupportsRevocationOptions())
+                    {
+                        if (revocationFlag == X509RevocationFlag.EndCertificateOnly)
+                        {
+                            // No way to specfiy end certificate only if revocation options are not available
+                            throw new PlatformNotSupportedException(SR.Format(SR.Chain_SettingNotSupported, $"{nameof(X509RevocationFlag)}.{nameof(X509RevocationFlag.EndCertificateOnly)}"));
+                        }
+
+                        // Defaults to offline when revocation options are not available
+                        if (revocationMode == X509RevocationMode.Online)
+                        {
+                            throw new NotImplementedException($"{nameof(Evaluate)} (X509RevocationMode.{revocationMode})");
+                        }
```
If the options aren't available then there's too much revocation checking?  Not enough?  The logic here suggests both things.

If the default is no revocation checking, then I think the right answer is to just report RevocationStatusUnknown.

If the default is something else I might have other opinions.

elinor-fung replied:
If the options aren't available, we can still turn revocation checking on or off. By default, it should be offline and exclude the root when turned on. From docs I've seen, we should still be able to enable/disable OCSP though setting the `ocsp.enable` security property: https://docs.oracle.com/javase/8/docs/technotes/guides/security/certpath/CertPathProgGuide.html#AppC
So I think the only thing that would truly be unsupported without the options is `X509RevocationFlag.EndCertificateOnly` (and `EntireChain`).

This PR only turns revocation checking on/off and doesn't handle online mode without the options API available (largely because I haven't set up my local environment with an version that doesn't have the options yet). My intent was to make a separate/follow-up change to remove this NotImplementedException and handle online mode in this case (assuming the property works the same on Android).

bartonjs replied:
I'm having trouble wading through the factory and such handwaviness in Java/Android's documentation here, but the notion that CRL/OCSP fetch matches our Offline mode seems weird.  (There's a cache, if it's in the cache, great, if not, you'll get RevocationStatusUnknown).

Our Online mode is "use the cache, if you can, but you're allowed to go download things if you have to".

bartonjs replied:
> My intent was to make a separate/follow-up change to remove this NotImplementedException and handle online mode in this case (assuming the property works the same on Android).

If that's the path forward, please create an issue to track the work, so that if you get pulled onto a different project we remember it's still pending :smile:

elinor-fung replied:
Rewrote https://github.com/dotnet/runtime/issues/45741 to be more reflective of the current direction (instead of the just P/Invokes one). I'm going with using that as the master work-tracking issue instead of multiple issues, since it is all this work is still on-going.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+            {
+                Debug.Assert(_chainContext != null);
+
+                long timeInMsFromUnixEpoch = new DateTimeOffset(verificationTime).ToUnixTimeMilliseconds();
```
I don't know which way Android revocation works, but would we do better with chain building if we always built the chain as of cert.NotBefore, then manually did the expired check ourselves afterwards?

bartonjs replied:
> I don't know which way Android revocation work

(Some libraries will return "not revoked" if the revocation time was in the future as compared to the verification time... some say "if it's on the list, it's revoked, I don't care about 'as of when'")

elinor-fung replied:
Using the `cert.NotBefore` for building the initial chain should let us actually build a chain when the verification time is outside the cert's validity and get some better chain building.

Would there be situations where a chain should pass time validation (so we could build it on Android) with the specified verification time, but not with `cert.NotBefore` - e.g. an issuer cert had a later `NotBefore` that the verification time satisfies, but not `cert.NotBefore`? That seems weird to me, but I don't know the area and didn't find anything saying that the issuer cert's valid time should fully contain the issued cert's valid time. (I do see that .NET has the `NotTimeNested` flag that the doc says is deprecated.)

elinor-fung replied:
Android revocation checking does care about 'as of when' and uses the verification time.

bartonjs replied:
> Android revocation checking does care about 'as of when' and uses the verification time.

😢 

So if we really cared to make expiration be soft-detected we'd need to do something like

* Check manually if it's expired
* Build with a valid time (e.g. the NotBefore time) and no revocation to get the full chain details
* For the newly chained-in certs, also manually check expiry (if the engine does nested validity time checking already then this can be skipped when the target cert wasn't expired).
* If revocation is requested
  * If expired: Slap RevocationStatusUnknown everywhere.
  * else: Use the requested time and build again with revocation on.

Sounds... complicated enough that I'm OK deferring it until complaints come in.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c:
```diff
@@ -0,0 +1,470 @@
+        (*env)->CallBooleanMethod(env, options, g_HashSetAdd, preferCrls);
+        (*env)->DeleteLocalRef(env, preferCrls);
+
+        // options.add(PKIXRevocationChecker.Option.NO_FALLBACK);
```
I don't know how the revocation works internally here (I haven't found a good description), but I'm not sure that PREFER_CRLS | NO_FALLBACK is quite the same as Offline.

Windows does:

1. Check the OCSP cache for a valid OCSP response.
2. If !offline, do an OCSP fetch
3. Check the CRL cache for a valid CRL entry.
4. If !offline, do a CRL fetch
5. No options remain, report RevocationStatusUnknown

Options 2 and 3 may be reversed, I'm not sure, it's a black box.

On Linux I think we're

1. CRL cache
2. OCSP cache
3. If !offline CRL fetch
4. If !offline OCSP fetch
5. RevocationStatusKnown

but I might be misremembering

macOS is similar, but they have "network is allowed" and "revocation is enabled" separate options.  The former controls both AIA/chain fetch and revocation fetch, so together we can have Offline mode.

If Android's CRL provider is really "you have to have already fetched it for me" but their OCSP provider is live, then I guess it matches, but then we're missing CRL+Online.

elinor-fung replied:
Ah, boo... Definitely not mapping this properly.

From what I've tested on an Android version where `PKIXRevocationChecker` is available:
- OCSP is live (when enabled)
- Using `PKIXRevocationChecker` (setting options, adding it as a checker), does CRL fetch
- Not using `PKIXRevocationChecker` (just using `setRevocationEnabled`), does not do CRL fetch

Haven't checked an older version yet.

I don't see anything for configuring the fetching. So I guess it would be that if we use `PKIXRevocationChecker`, we are always Online and not using it, we are missing CRL+Online and Offline can't have OCSP.

bartonjs replied:
If Offline gets treated as Online, that's probably not the end of the world.  Where do the CRLs come from in "Not using PKIXRevocationChecker"?  If it's from an internal cache, then I think the mapping is

* Online: OCSP enabled, use PKIXRevocationChecker
* Offline: OCSP disabled, just setRevocationEnabled(true)
* NoCheck: setRevocationEnabled(false)

If it's "you have to give it to us" then perhaps just do Offline => Online for now with a vWhenever issue for properly respecting Offline mode.  (Presumably you can get the CRLs off of the chain to save for later, and then we'd just have to squirrel them away somewhere...)

elinor-fung replied:
Looks like checking an internal cache is wrapped up in using `PKIXRevocationChecker` and when not using that, it is 'you have to give it to us'.

elinor-fung replied:
Treating `Offline` as `Online` now (with a debug log message).

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs:
```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+
+                    if (!Interop.AndroidCrypto.X509ChainSupportsRevocationOptions())
+                    {
+                        if (revocationFlag == X509RevocationFlag.EndCertificateOnly)
+                        {
+                            // No way to specfiy end certificate only if revocation options are not available
+                            throw new PlatformNotSupportedException(SR.Format(SR.Chain_SettingNotSupported, $"{nameof(X509RevocationFlag)}.{nameof(X509RevocationFlag.EndCertificateOnly)}"));
+                        }
```
Do we get to learn what level was revoked?  If so, the fallback would be "build the chain with revocation on, if revocation occurred past the end-entity certificate, build it again with revocation off, if there are no errors, everything's cool".  (Or, if revocation is guaranteed to be the last thing reported we could just ignore it at levels beyond the end-entity cert)

elinor-fung replied:
We do get the index, but the validation throws at the first error and the checker does its checks from most-trusted to target, so we don't know that the end cert actually passes the revocation check if an intermediate failed the revocation check.

bartonjs replied:
Oh, yep.  Right.  Build from bottom up, check revocation from top down (because if the middle is revoked then you can't actually trust the CRL that says the next level isn't revoked).

While there are some tests that care, generally specifying EndCertificateOnly is just about a perf win.  On the versions that don't support it I think we can just let all three levels be evaluated as ExcludeRoot.  (Basically, consider it as "we set the option, but the underlying implementation ignored it"... we did the best we could.  Not a huge problem because we're failing in the safer direction.)

elinor-fung replied:
Treating `EndCertificateOnly` as `ExcludeRoot` now when we don't have the options.

elinor-fung committed:
Merge remote-tracking branch 'upstream/main' into chainPal-android

elinor-fung committed:
Stop throwing PNSE on verification/revocation flags that aren't respected

Treat X509RevocationFlag EntireChain as ExcludeRoot
If revocation options aren't available, treat X509RevocationFlag EndCertificateOnly as ExcludeRoot

elinor-fung committed:
Update tests based on lack of AIA fetching support

elinor-fung committed:
Treat X509RevocationMode Offline as Online

Try to get better revocation status when possible by retrying validation without revocation

elinor-fung committed:
Make test revocation responder handle POST requests for OCSP

elinor-fung committed:
Handle EndCertificateOnly revocation check with root-only chain

Validation in two passes - with / without revocation for better status

elinor-fung committed:
AddRef cert handles

elinor-fung committed:
Update DynamicRevocationTests for Android behaviour

elinor-fung commented:
Okay. PR description has been updated. With the latest updates, the chain-related tests should be passing except for `CreateChain_RSAPSS`.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c:
```diff
@@ -0,0 +1,598 @@
+        }
+    }
+
+    return PAL_X509ChainPartialChain;
```
Can we do better than this?

The main thing is that we never return `true` from chain.Build if we don't know why something failed, so this is sort of OK, but I know I'm about to ask why a PartialChain is popping up and I'm guessing it's this fallback here.

On macOS our details processing comes from weak string processing, and anytime we get an unknown value we throw an exception (since we don't know if unknown means fatal or innocuous).

elinor-fung replied:
A big source of `PartialChain` without any details would be from the initial build (instead of the second-pass validation). Unfortunately, all seem to come as a `CertPathBuilderException` with a generic message (`unable to find valid certification path to requested target`) and no wrapped exception, so I don't think there's any better we can do in that case.

In the case of the `CertPathValidatorException` without a reason (not set or on a version where it doesn't exist), I think we could do something like walk through any wrapped exceptions and check for specific exception types (e.g. the subclasses of [`CertificateException`](https://developer.android.com/reference/java/security/cert/CertificateException)) to determine a better status.

Testing on a version that has the reason API available, I was seeing errors with it set such that we could map them to a status, so I think it is on older versions where trying additional metrics would be helpful. I'll add it as an item to  #45741.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c:
```diff
@@ -0,0 +1,598 @@
+            jobject reason = (*env)->CallObjectMethod(env, error, g_CertPathValidatorExceptionGetReason);
+            chainStatus = ChainStatusFromValidatorExceptionReason(env, reason);
+            (*env)->DeleteLocalRef(env, reason);
+        }
```
Shouldn't this `else` set chainStatus to... something?

We have a validation exception, but can't ask what went wrong... that's not "NoError" :)

elinor-fung replied:
Err, yep. Apparently, have the managed just forcing `NoError` -> `PartialChain` for... reasons? Will make this fallback, since that makes way more sense.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c:
```diff
@@ -0,0 +1,598 @@
+    loc[result] =
+        (*env)->CallObjectMethod(env, loc[validator], g_CertPathValidatorValidate, ctx->certPath, ctx->params);
+    bool hadErrorsWithoutRevocation = TryGetJNIException(env, &loc[ex], false /*printException*/);
+    if (hadErrorsWithoutRevocation)
+    {
+        (*env)->CallBooleanMethod(env, ctx->errorList, g_ArrayListAdd, loc[ex]);
+    }
+
+    if (revocationMode != X509RevocationMode_NoCheck)
```
If there were errors without revocation is there any chance of success with revocation? (My gut says no)

If there isn't, then it seems like the revocationMode check should be `else if`, not `if`, and then the redundant error code copy inside `ValidateWithRevocation` can be removed.

But I could just be missing something :smile:.

elinor-fung replied:
No chance, indeed. Will avoid `ValidateWithRevocation` if there were errors and put `RevocationStatusUnknown` on everything if checking was requested.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:
```diff
@@ -17,11 +17,33 @@ public static class DynamicRevocationTests
+        // after will have PartialChain and RevocationStatusUnknown instead of Revoked
+        private static readonly X509ChainStatusFlags ThisOsRevokedWithPreviousRevocationError =
+                PlatformDetection.IsAndroid ?
+                X509ChainStatusFlags.PartialChain | X509ChainStatusFlags.RevocationStatusUnknown :
```
This doesn't make sense to me... If we know something got revoked, why is PartialChain popping up? We built a full chain...

elinor-fung replied:
Ah, yeah, this was from always setting `PartialChain` after any error. Should be able to fix this, since I have with / without revocation split now.

elinor-fung committed:
Merge branch 'main' into chainPal-android

elinor-fung committed:
Skip validation with revocation if without revocation failed

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -269,11 +289,22 @@ public enum BuildChainCustomTrustStoreTestArguments : int
-        [InlineData(true, X509ChainStatusFlags.NoError, BuildChainCustomTrustStoreTestArguments.UntrustedIntermediateTrustedRoot)]
-        [InlineData(true, X509ChainStatusFlags.NoError, BuildChainCustomTrustStoreTestArguments.TrustedIntermediateTrustedRoot)]
-        [InlineData(true, X509ChainStatusFlags.NoError, BuildChainCustomTrustStoreTestArguments.MultipleCalls)]
+        [PlatformSpecific(~TestPlatforms.Android)] // Test relies on AIA fetching, which Android does not support
```
The AIA fetch and system retrieval aren't really important to this test.  Making all platforms bring the data in explicitly is fine.

IIRC, it was just written before we bothered having all of the certs in this chain explicitly... so it was "eh, we already have a way of getting at this data".

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -441,54 +504,73 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
 
                 // Ignore anything except NotTimeValid
                 chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllFlags & ~X509VerificationFlags.IgnoreNotTimeValid;
+
+                // Android doesn't support certain flags and will error if the chain is
+                // invalid and one of those flags is set.
+                if (PlatformDetection.IsAndroid)
+                {
+                    chain.ChainPolicy.VerificationFlags &= ~X509VerificationFlags.AllowUnknownCertificateAuthority;
+                    chain.ChainPolicy.VerificationFlags &= ~X509VerificationFlags.IgnoreInvalidName;
+                    chain.ChainPolicy.VerificationFlags &= ~X509VerificationFlags.IgnoreInvalidPolicy;
+                }
+
```
I don't think this is true any longer?

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -950,6 +1041,43 @@ void CheckChain()
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Android)]
+        public static void BuildInvalidSignatureTwice_Android()
```
Like some prior tests, the AIA fetch/etc during this test weren't really part of the test.  If you want to merge the two back together that'd probably be better.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:
```diff
@@ -79,10 +94,22 @@ public static IEnumerable<object[]> AllViableRevocation
         [MemberData(nameof(AllViableRevocation))]
         public static void NothingRevoked(PkiOptions pkiOptions)
         {
+            bool usingCrl = pkiOptions.HasFlag(PkiOptions.IssuerRevocationViaCrl) || pkiOptions.HasFlag(PkiOptions.EndEntityRevocationViaCrl);// && !pkiOptions.HasFlag(PkiOptions.EndEntityRevocationViaOcsp);
```
Please delete the commented out code.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:
```diff
@@ -1248,14 +1372,21 @@ public static void TestRevocation_Offline_Revoked()
+            if (PlatformDetection.IsAndroid)
+            {
+                // Android stops validation at the first failure, so the end certificate would
+                // end up marked with PartialChain and RevocationStatusUnknown
```
I believe the PartialChain part is out of date.

bartonjs commented:
Alrighty, I think we're good on the product code, but a few bits of evolutionary cleanup remain in the test code.  Almost done!

elinor-fung committed:
Update tests

elinor-fung committed:
Remove unnecessary dispose

elinor-fung commented:
/backport to release/6.0-preview3
